File: fsharp/GildedRose/GildedRose.fs
module GildedRose

open System.Collections.Generic

type Item = { Name: string; SellIn: int; Quality: int }

type GildedRose(items:IList<Item>) as this =
    let Items = items

    member this.UpdateQuality() =
        for i = 0 to Items.Count - 1 do
            if Items.[i].Name <> "Aged Brie" && Items.[i].Name <> "Backstage passes to a TAFKAL80ETC concert" then
                if Items.[i].Quality > 0 then
                    if Items.[i].Name <> "Sulfuras, Hand of Ragnaros" then
                        Items.[i] <- { Items.[i] with Quality = (Items.[i].Quality - 1) } 
            else
               if Items.[i].Quality < 50 then
                    Items.[i] <- { Items.[i] with Quality = (Items.[i].Quality + 1) } 
                    if Items.[i].Name = "Backstage passes to a TAFKAL80ETC concert" then
                        if Items.[i].SellIn < 11 then
                            if Items.[i].Quality < 50 then
                                Items.[i] <- { Items.[i] with Quality = (Items.[i].Quality + 1) } 
                        if Items.[i].SellIn < 6 then
                            if Items.[i].Quality < 50 then
                                Items.[i] <- { Items.[i] with Quality = (Items.[i].Quality + 1) } 
            if Items.[i].Name <> "Sulfuras, Hand of Ragnaros" then                 
                Items.[i] <- { Items.[i] with SellIn  = (Items.[i].SellIn - 1) } 
            if Items.[i].SellIn < 0 then
                if Items.[i].Name <> "Aged Brie" then
                    if Items.[i].Name <> "Backstage passes to a TAFKAL80ETC concert" then
                        if Items.[i].Quality > 0 then
                            if Items.[i].Name <> "Sulfuras, Hand of Ragnaros" then
                                Items.[i] <- { Items.[i] with Quality   = (Items.[i].Quality  - 1) } 
                    else
                        Items.[i] <- { Items.[i] with Quality   = (Items.[i].Quality  - Items.[i].Quality) } 
                else
                    if Items.[i].Quality < 50 then
                        Items.[i] <- { Items.[i] with Quality   = (Items.[i].Quality + 1) }  
        ()

[<EntryPoint>]
let main argv = 
    printfn "OMGHAI!"
    let Items = new List<Item>()
    Items.Add({Name = "+5 Dexterity Vest"; SellIn = 10; Quality = 20})
    Items.Add({Name = "Aged Brie"; SellIn = 2; Quality = 0})
    Items.Add({Name = "Elixir of the Mongoose"; SellIn = 5; Quality = 7})
    Items.Add({Name = "Sulfuras, Hand of Ragnaros"; SellIn = 0; Quality = 80})
    Items.Add({Name = "Sulfuras, Hand of Ragnaros"; SellIn = -1; Quality = 80})
    Items.Add({Name = "Backstage passes to a TAFKAL80ETC concert"; SellIn = 15; Quality = 20})
    Items.Add({Name = "Backstage passes to a TAFKAL80ETC concert"; SellIn = 10; Quality = 49})
    Items.Add({Name = "Backstage passes to a TAFKAL80ETC concert"; SellIn = 5; Quality = 49})
    Items.Add({Name = "Conjured Mana Cake"; SellIn = 3; Quality = 6})

    let app = new GildedRose(Items)
    for i = 0 to 30 do
        printfn "-------- day %d --------" i
        printfn "name, sellIn, quality"
        for j = 0 to Items.Count - 1 do
             printfn "%s, %d, %d" Items.[j].Name Items.[j].SellIn Items.[j].Quality
        printfn ""
        app.UpdateQuality()
    0 

File: go/gildedrose/gildedrose.go
package gildedrose

import (
	"github.com/emilybache/gildedrose-refactoring-kata/gildedrose/domain"
)

const (
	AgedBrie            = "Aged Brie"
	BackstagePasses     = "Backstage passes to a TAFKAL80ETC concert"
	Sulfuras            = "Sulfuras, Hand of Ragnaros"
	Conjured            = "Conjured Mana Cake"
	MaxQuality          = 50
	MinQuality          = 0
	SulfurasQuality     = 80
	Expired             = 0
	Backstage10Days     = 10
	Backstage5Days      = 5
	NormalDegradeRate   = 1
	ConjuredDegradeRate = 2
)

func UpdateQuality(items []*domain.Item) {
	for _, item := range items {
		updateItem(item)
		validateQuality(item)
	}
}
func validateQuality(item *domain.Item) {
	if item.Name == Sulfuras {
		if item.Quality != SulfurasQuality {
			item.Quality = SulfurasQuality
		}
		return
	}
	if item.Quality > MaxQuality || item.Quality < MinQuality {
		item.Quality = MinQuality
	}
}
func updateItem(item *domain.Item) {
	switch item.Name {
	case AgedBrie:
		incrementQuality(item)
	case BackstagePasses:
		updateBackstagePasses(item)
	case Sulfuras:
		// Sulfuras does not change in quality or sell-in
	case Conjured:
		degradeQuality(item, ConjuredDegradeRate)
	default:
		degradeQuality(item, NormalDegradeRate)
	}

	if item.Name != Sulfuras {
		item.SellIn--
	}

	if item.SellIn < Expired {
		handleExpiredItem(item)
	}
}

func incrementQuality(item *domain.Item) {
	if item.Quality < MaxQuality {
		item.Quality++
	}
}
func updateBackstagePasses(item *domain.Item) {
	incrementQuality(item)
	if item.SellIn <= Backstage10Days && item.Quality < MaxQuality {
		item.Quality++
	}
	if item.SellIn <= Backstage5Days && item.Quality < MaxQuality {
		item.Quality++
	}
}

func degradeQuality(item *domain.Item, degradeRate int) {
	if item.Quality > MinQuality {
		item.Quality -= degradeRate
	}
}

func handleExpiredItem(item *domain.Item) {
	switch item.Name {
	case AgedBrie:
		incrementQuality(item)
	case BackstagePasses:
		item.Quality = MinQuality
	case Conjured:
		degradeQuality(item, ConjuredDegradeRate)
	case Sulfuras:
		// Sulfuras does not change in quality or sell-in
	default:
		degradeQuality(item, NormalDegradeRate)
	}
}
